### PR TITLE
fast_concat() for Metric

### PR DESCRIPTION
**Pull request by kockan**
Intended for #46

### CODE REVIEW

**Review by TedBrookings — CHANGES_REQUESTED** [review]
This is a good submission and there's nothing strictly _wrong_ with it, but there are a few minor changes I'd like to see before merging.
1) I think that overall there would be less code maintenance if the specialized appending functions were instead rolled into the writer functions with an `append` input defaulting to False. If you disagree I'm willing to entertain push-back.
2) I want to see more detailed messages when the headers aren't correct.

File: fgpyo/io/__init__.py [hunk 1]
```diff
@@ -157,6 +157,21 @@ def to_writer(path: Path) -> Union[IO[Any], io.TextIOWrapper]:
...
+
+    Example:
+        appender = fio.to_appender(path = Path("appender.txt"))
+        appender.write(f'{something}\n')
+        appender.close()
+    """
+    if path.suffix in COMPRESSED_FILE_EXTENSIONS:
+        return io.TextIOWrapper(cast(IO[bytes], gzip.open(path, mode="ab")), encoding="utf-8")
+    else:
+        return path.open(mode="a")
```
[the remark this answers is not in the document]
Comment (kockan) [reply fgpyo/io/__init__.py]: Done.

File: fgpyo/io/__init__.py [hunk 2]
```diff
@@ -196,3 +210,20 @@ def write_lines(path: Path, lines_to_write: Iterable[Any]) -> None:
...
+
+    Example:
+        lines: List[Any] = ["things to append", 100]
+        path_to_append_to: Path = Path("file_to_append_to.txt")
+        fio.append_lines(path = path_to_append_to, lines_to_append = lines)
+    """
+    with to_appender(path=path) as appender:
+        for line in lines_to_append:
+            appender.write(str(line))
+            appender.write("\n")
```
[the remark this answers is not in the document]
Comment (kockan) [reply fgpyo/io/__init__.py]: Done.

File: fgpyo/util/metric.py [hunk 3]
```diff
@@ -305,3 +305,17 @@ def format_value(cls, value: Any) -> str:
...
         return [] if value == "" else value.split(",")
+
+    @staticmethod
+    def fast_concat(*inputs: Path, output: Path) -> None:
+        headers = []
+        for input_path in inputs:
+            header = next(io.read_lines(input_path))
+            headers.append(header)
+
+        assert len(headers) > 0
```
[the remark this answers is not in the document]
Comment (kockan) [reply fgpyo/util/metric.py]: Done.

File: fgpyo/util/metric.py [hunk 4]
```diff
@@ -305,3 +305,17 @@ def format_value(cls, value: Any) -> str:
...
+        assert len(set(headers)) == 1
+
+        io.write_lines(path = output, lines_to_write = set(headers))
+        for input_path in inputs:
+            io.append_lines(path = output, lines_to_append = list(io.read_lines(input_path))[1:])
```
[the remark this answers is not in the document]
Comment (kockan) [reply fgpyo/util/metric.py]: Done.

File: fgpyo/util/metric.py [hunk 5]
```diff
@@ -305,3 +305,17 @@ def format_value(cls, value: Any) -> str:
...
+
+    @staticmethod
+    def fast_concat(*inputs: Path, output: Path) -> None:
+        headers = []
+        for input_path in inputs:
+            header = next(io.read_lines(input_path))
+            headers.append(header)
+
+        assert len(headers) > 0
+        assert len(set(headers)) == 1
```
[the remark this answers is not in the document]
Comment (kockan) [reply fgpyo/util/metric.py]: Done.

**Comment by kockan** [comment]
I don't believe there's anything I disagree with in this review. Looks much more elegant now. Thanks for the suggestions! Let me know if this needs any further edits.

**Comment by TedBrookings** [comment]
It's nearly there, but it looks like there's a few style issues. I think you need to run `ci/check.sh`. I'm not sure why the github hooks didn't run automatically for that. Other than that it's ready.

**Review by TedBrookings — APPROVED** [review]
Looks great now, thanks for your contribution!
Be sure to "Squash and merge" vs the other options.

**Comment by tfenne** [comment]
@TedBrookings I don't think @kockan has write permissions so you'll need to do the merge!

**Pull request merged**